pkg/vfs/vfs.go

```diff
@@ -0,0 +1,283 @@
+		return child.Create(otherElements)
+	}
+
+	// We do not yet have an overlay INode, create one and
+	// continue downward.
+	if i.Children == nil {
+		i.Children = make(map[string]*INode)
+	}
+
+	child := &INode{Filename: currentElement}
+	i.Children[currentElement] = child
+	return child.Create(otherElements)
+}
+
+func (i *INode) walkOrCreate(path string) (*INode, error) {
+	node, err := i.walk(path)
+	if err != nil {
+		// No overlay node, create a new one.
+		node, err = i.Create(path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return node, nil
+}
+
+// Chmod updates the permissions on an INode.
+func (i *INode) Chmod(path string, mode fs.FileMode) error {
+	node, err := i.walkOrCreate(path)
+	if err != nil {
+		return err
+	}
+
+	isDir := node.IsDir()
+	newMode := mode
+
+	if isDir {
+		newMode |= fs.ModeDir
+	}
+
+	node.UnderlayINode.Mode = uint32(newMode)
+	return nil
+}
+
+// Chown updates the ownership on an INode.
+func (i *INode) Chown(path string, uid, gid uint32) error {
+	node, err := i.walkOrCreate(path)
+	if err != nil {
+		return err
+	}
+
+	node.UnderlayINode.Uid = uid
+	node.UnderlayINode.Gid = gid
+
+	return nil
+}
+
+func (i INode) IsDir() bool {
+	return i.Mode()&fs.ModeDir == fs.ModeDir
+}
+
+func (i INode) Mode() fs.FileMode {
+	return fs.FileMode(i.UnderlayINode.Mode)
+}
+
+func (i INode) ModTime() time.Time {
+	ts, tns := i.UnderlayINode.Mtim.Unix()
+	return time.Unix(ts, tns)
+}
+
+func (i INode) Name() string {
+	return i.Filename
+}
+
+func (i INode) Size() int64 {
+	return i.UnderlayINode.Size
+}
+
+func (i INode) Sys() any {
+	return &i.UnderlayINode
+}
+
+func (i INode) Info() (os.FileInfo, error) {
+	return i, nil
+}
+
+func (i INode) Type() fs.FileMode {
+	return i.Mode()
+}
+
+// A VFS is an overlay virtual filesystem which tracks an underlying
+// io/fs.FS.
+type BaseFS interface {
+	fs.FS
+	fs.ReadDirFS
+	fs.ReadFileFS
+	fs.StatFS
+
+	Create(path string) (io.WriteCloser, error)
+	Remove(path string) error
+	RemoveAll(path string) error
+}
+
+type VFS struct {
+	FS   BaseFS
+	Root *INode
+}
+
+func (vfs *VFS) Stat(path string) (os.FileInfo, error) {
+	inode, err := vfs.Root.Stat(path)
+	if err == nil {
+		return inode, err
+	}
+
+	return vfs.FS.Stat(path)
+}
+
+func (vfs *VFS) Create(path string) (io.WriteCloser, error) {
+	return vfs.FS.Create(path)
+}
+
+func (vfs *VFS) Open(path string) (fs.File, error) {
+	return vfs.FS.Open(path)
+}
+
+func (vfs *VFS) ReadFile(path string) ([]byte, error) {
+	return vfs.FS.ReadFile(path)
+}
+
+func (vfs *VFS) ReadDir(path string) ([]fs.DirEntry, error) {
+	de, err := vfs.FS.ReadDir(path)
+	if err != nil {
+		return []fs.DirEntry{}, err
+	}
+
+	baseINode, err := vfs.Root.walk(path)
+	if err != nil {
+		return de, nil
+	}
+
+	out := []fs.DirEntry{}
+	for _, dentry := range de {
+		if patchedINode, ok := baseINode.Children[dentry.Name()]; ok {
+			out = append(out, patchedINode)
+		} else {
+			out = append(out, dentry)
+		}
+	}
```

Comment:
So this is the core of the VFS, right? I'll try to regurgitate this to see if I'm getting it:

The way I understand the code is that the VFS will return the calls chrooted to the underlying directory. But I can create "phantom" paths (handled with `iNode`) which don't really exist in the underlying directory. Those phantom paths take precedence over the real files in the source directory. 

Is that a fair interpretation?

Reply:
well, the file content is backed by the underlying directory, but the permissions / ownership are phantom :)